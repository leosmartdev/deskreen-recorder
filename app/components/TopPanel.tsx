/* eslint-disable react/destructuring-assignment */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react-hooks/rules-of-hooks */
import { shell } from 'electron';
import React, { useCallback, useContext } from 'react';
import { Button, Text, Icon, Position, Tooltip } from '@blueprintjs/core';
import { createStyles, makeStyles } from '@material-ui/core/styles';
import { Col, Row } from 'react-flexbox-grid';
import { useTranslation } from 'react-i18next';
import SettingsOverlay from './SettingsOverlay/SettingsOverlay';
import ConnectedDevicesListDrawer from './ConnectedDevicesListDrawer';
import { SettingsContext } from '../containers/SettingsProvider';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore fine import here
import RedHeartTwemojiPNG from '../images/red_heart_2764_twemoji_120x120.png';

const useStylesWithTheme = (isDarkTheme: boolean) =>
  makeStyles(() =>
    createStyles({
      topPanelRoot: {
        display: 'flex',
        paddingTop: '15px',
        marginBottom: '20px',
      },
      logoWithAppName: { margin: '0 auto' },
      appNameHeader: {
        margin: '0 auto',
        paddingTop: '5px',
        fontFamily: 'Lexend Peta',
        fontSize: '20px',
        color: isDarkTheme ? '#48AFF0' : '#1F4B99',
        cursor: 'default !important',
      },
      topPanelControlButtonsRoot: {
        position: 'absolute',
        right: '15px',
        display: 'flex',
      },
      topPanelControlButton: {
        width: '40px',
        height: '40px',
        borderRadius: '50px',
        cursor: 'default !important',
      },
      topPanelControlButtonMargin: {
        marginRight: '20px',
        cursor: 'default !important',
      },
      topPanelIconOfControlButton: {
        cursor: 'default !important',
      },
    })
  );

export default function TopPanel(props: any) {
  const { t } = useTranslation();
  const { isDarkTheme } = useContext(SettingsContext);

  const getClassesCallback = useCallback(() => {
    return useStylesWithTheme(isDarkTheme)();
  }, [isDarkTheme]);

  const [isSettingsOpen, setIsSettingsOpen] = React.useState(false);
  const [isDrawersOpen, setIsDrawerOpen] = React.useState(false);

  const handleSettingsOpen = useCallback(() => {
    setIsSettingsOpen(true);
  }, []);

  const handleSettingsClose = useCallback(() => {
    setIsSettingsOpen(false);
  }, []);

  const handleToggleConnectedDevicesListDrawer = useCallback(() => {
    setIsDrawerOpen(!isDrawersOpen);
  }, [isDrawersOpen]);

  const donateTooltipContent = t(
    'If you like Deskreen consider contributing financially Deskreen is open-source Your donations keep us motivated to make Deskreen even better'
  );
  const deskreenButtonTooltip = t('Click to visit our website');

  const renderDonateButton = useCallback(() => {
    return (
      <Tooltip content={donateTooltipContent} position={Position.BOTTOM}>
        <Button
          style={{
            marginRight: '10px',
            borderRadius: '100px',
          }}
          onClick={() => {
            shell.openExternal('https://www.patreon.com/deskreen');
          }}
        >
          <Row start="xs">
            <Col xs>
              <img
                src={RedHeartTwemojiPNG}
                width={16}
                height={16}
                style={{ transform: 'translateY(2px)' }}
                alt="heart"
              />
            </Col>
            <Col xs>
              <div
                style={{
                  transform: 'translateY(2px) translateX(-5px)',
                  width: 'max-content',
                }}
              >
                <Text>{t('Donate')}</Text>
              </div>
            </Col>
          </Row>
        </Button>
      </Tooltip>
    );
  }, [donateTooltipContent, t]);

  const renderConnectedDevicesListButton = useCallback(() => {
    return (
      <div className={getClassesCallback().topPanelControlButtonMargin}>
        <Tooltip content={t('Connected Devices')} position={Position.BOTTOM}>
          <Button
            id="top-panel-connected-devices-list-button"
            intent="primary"
            className={getClassesCallback().topPanelControlButton}
            onClick={handleToggleConnectedDevicesListDrawer}
          >
            <Icon
              className={getClassesCallback().topPanelIconOfControlButton}
              icon="th-list"
              iconSize={20}
            />
          </Button>
        </Tooltip>
      </div>
    );
  }, [getClassesCallback, handleToggleConnectedDevicesListDrawer, t]);

  const renderHelpButton = useCallback(() => {
    return (
      <div className={getClassesCallback().topPanelControlButtonMargin}>
        <Tooltip content={t('Tutorial')} position={Position.BOTTOM}>
          <Button
            id="top-panel-help-button"
            intent="none"
            className={getClassesCallback().topPanelControlButton}
            onClick={() => {
              shell.openExternal('https://www.deskreen.com/?#howtos');
            }}
          >
            <Icon
              className={getClassesCallback().topPanelIconOfControlButton}
              icon="learning"
              iconSize={22}
            />
          </Button>
        </Tooltip>
      </div>
    );
  }, [getClassesCallback, t]);

  const renderSettingsButton = useCallback(() => {
    return (
      <div className={getClassesCallback().topPanelControlButtonMargin}>
        <Tooltip content={t('Settings')} position={Position.BOTTOM}>
          <Button
            id="top-panel-settings-button"
            onClick={handleSettingsOpen}
            className={getClassesCallback().topPanelControlButton}
          >
            <Icon
              className={getClassesCallback().topPanelIconOfControlButton}
              icon="cog"
              iconSize={22}
            />
          </Button>
        </Tooltip>
      </div>
    );
  }, [getClassesCallback, handleSettingsOpen, t]);

  const renderLogoWithAppName = useCallback(() => {
    return (
      <div
        id="logo-with-popover-visit-website"
        className={getClassesCallback().logoWithAppName}
      >
        <Tooltip content={deskreenButtonTooltip} position={Position.BOTTOM}>
          <Button
            minimal
            onClick={() => {
              shell.openExternal('https://www.deskreen.com');
            }}
            style={{
              borderRadius: '100px',
            }}
          >
            <h4
              id="deskreen-top-app-name-header"
              className={getClassesCallback().appNameHeader}
              style={{
                transform: 'translateY(-3px)',
              }}
            >
              Deskreen
            </h4>
          </Button>
        </Tooltip>
      </div>
    );
  }, [deskreenButtonTooltip, getClassesCallback]);

  return (
    <>
      <div className={getClassesCallback().topPanelRoot}>
        <Row
          middle="xs"
          center="xs"
          style={{ width: '100%', transform: 'translateX(-50px)' }}
        >
          <Col>{renderDonateButton()}</Col>
          <Col>{renderLogoWithAppName()}</Col>
        </Row>
        <div className={getClassesCallback().topPanelControlButtonsRoot}>
          {renderConnectedDevicesListButton()}
          {renderHelpButton()}
          {renderSettingsButton()}
        </div>
      </div>
      <SettingsOverlay
        isSettingsOpen={isSettingsOpen}
        handleClose={handleSettingsClose}
      />
      <ConnectedDevicesListDrawer
        isOpen={isDrawersOpen}
        handleToggle={handleToggleConnectedDevicesListDrawer}
        stepperRef={props.stepperRef}
      />
    </>
  );
}
