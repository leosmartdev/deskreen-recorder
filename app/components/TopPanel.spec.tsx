/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import Enzyme, { ShallowWrapper } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import TopPanel from './TopPanel';

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();

const settingsButtonSelector = '#top-panel-settings-button';
const connectedDevicesListButtonSelector =
  '#top-panel-connected-devices-list-button';

describe('<TopPanel />', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = Enzyme.shallow(<TopPanel />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match exact snapshot', () => {
    expect(EnzymeToJson(wrapper)).toMatchSnapshot();
  });

  describe('when Settings button in top panel is clicked', () => {
    it('should Settings Overlay props isSettingsOpen to true', () => {
      // @ts-ignore
      wrapper.find(settingsButtonSelector).props().onClick();
      // @ts-ignore
      const settingsOverlay = wrapper.props().children[1];

      expect(settingsOverlay.props.isSettingsOpen).toBe(true);
    });
  });

  describe('when Settings panel is opened and when its props.handleClose is called', () => {
    it('should Settings Overlay props isSettingsOpen to false', () => {
      // @ts-ignore
      wrapper.find(settingsButtonSelector).props().onClick();
      // @ts-ignore
      const settingsOverlay = wrapper.props().children[1];
      settingsOverlay.props.handleClose();
      // @ts-ignore
      const settingsOverlayAfterClose = wrapper.props().children[1];

      expect(settingsOverlayAfterClose.props.isSettingsOpen).toBe(false);
    });
  });

  describe('when Connected Devices button in top panel is clicked', () => {
    it('should set Connected Devices List Drawer props isOpen to true', () => {
      // @ts-ignore
      wrapper.find(connectedDevicesListButtonSelector).props().onClick();
      // @ts-ignore
      const connectedDevicesListDrawer = wrapper.props().children[2];
      expect(connectedDevicesListDrawer.props.isOpen).toBe(true);
    });
  });

  describe("when Connected Devices List Drawer is opened, and when it's props.handleToggle is called", () => {
    it('should set Connected Devices List Drawer props isOpen to false', () => {
      // @ts-ignore
      wrapper.find(connectedDevicesListButtonSelector).props().onClick();
      // @ts-ignore
      const connectedDevicesListDrawer = wrapper.props().children[2];
      connectedDevicesListDrawer.props.handleToggle();
      // @ts-ignore
      const connectedDevicesListDrawerAfter = wrapper.props().children[2];

      expect(connectedDevicesListDrawerAfter.props.isOpen).toBe(false);
    });
  });
});
