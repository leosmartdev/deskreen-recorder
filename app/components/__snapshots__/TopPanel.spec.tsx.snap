// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopPanel /> should match exact snapshot 1`] = `
<Fragment>
  <div
    className="makeStyles-topPanelRoot-1"
  >
    <Row
      center="xs"
      middle="xs"
      style={
        Object {
          "transform": "translateX(-50px)",
          "width": "100%",
        }
      }
    >
      <Col>
        <Blueprint3.Tooltip
          content=""
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          minimal={false}
          position="bottom"
          transitionDuration={100}
        >
          <Blueprint3.Button
            onClick={[Function]}
            style={
              Object {
                "borderRadius": "100px",
                "marginRight": "10px",
              }
            }
          >
            <Row
              start="xs"
            >
              <Col
                xs={true}
              >
                <img
                  alt="heart"
                  height={16}
                  src="test-file-stub"
                  style={
                    Object {
                      "transform": "translateY(2px)",
                    }
                  }
                  width={16}
                />
              </Col>
              <Col
                xs={true}
              >
                <div
                  style={
                    Object {
                      "transform": "translateY(2px) translateX(-5px)",
                      "width": "max-content",
                    }
                  }
                >
                  <Blueprint3.Text
                    ellipsize={false}
                    tagName="div"
                  />
                </div>
              </Col>
            </Row>
          </Blueprint3.Button>
        </Blueprint3.Tooltip>
      </Col>
      <Col>
        <div
          className="makeStyles-logoWithAppName-9"
          id="logo-with-popover-visit-website"
        >
          <Blueprint3.Tooltip
            content=""
            hoverCloseDelay={0}
            hoverOpenDelay={100}
            minimal={false}
            position="bottom"
            transitionDuration={100}
          >
            <Blueprint3.Button
              minimal={true}
              onClick={[Function]}
              style={
                Object {
                  "borderRadius": "100px",
                }
              }
            >
              <h4
                className="makeStyles-appNameHeader-17"
                id="deskreen-top-app-name-header"
                style={
                  Object {
                    "transform": "translateY(-3px)",
                  }
                }
              >
                Deskreen
              </h4>
            </Blueprint3.Button>
          </Blueprint3.Tooltip>
        </div>
      </Col>
    </Row>
    <div
      className="makeStyles-topPanelControlButtonsRoot-25"
    >
      <div
        className="makeStyles-topPanelControlButtonMargin-34"
      >
        <Blueprint3.Tooltip
          content=""
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          minimal={false}
          position="bottom"
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="makeStyles-topPanelControlButton-40"
            id="top-panel-connected-devices-list-button"
            intent="primary"
            onClick={[Function]}
          >
            <Blueprint3.Icon
              className="makeStyles-topPanelIconOfControlButton-49"
              icon="th-list"
              iconSize={20}
            />
          </Blueprint3.Button>
        </Blueprint3.Tooltip>
      </div>
      <div
        className="makeStyles-topPanelControlButtonMargin-55"
      >
        <Blueprint3.Tooltip
          content=""
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          minimal={false}
          position="bottom"
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="makeStyles-topPanelControlButton-61"
            id="top-panel-help-button"
            intent="none"
            onClick={[Function]}
          >
            <Blueprint3.Icon
              className="makeStyles-topPanelIconOfControlButton-70"
              icon="learning"
              iconSize={22}
            />
          </Blueprint3.Button>
        </Blueprint3.Tooltip>
      </div>
      <div
        className="makeStyles-topPanelControlButtonMargin-76"
      >
        <Blueprint3.Tooltip
          content=""
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          minimal={false}
          position="bottom"
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="makeStyles-topPanelControlButton-82"
            id="top-panel-settings-button"
            onClick={[Function]}
          >
            <Blueprint3.Icon
              className="makeStyles-topPanelIconOfControlButton-91"
              icon="cog"
              iconSize={22}
            />
          </Blueprint3.Button>
        </Blueprint3.Tooltip>
      </div>
    </div>
  </div>
  <SettingsOverlay
    handleClose={[Function]}
    isSettingsOpen={false}
  />
  <ConnectedDevicesListDrawer
    handleToggle={[Function]}
    isOpen={false}
  />
</Fragment>
`;
